ocaml/xapi-idl/lib/posix_channel.ml

```diff
@@ -153,27 +167,40 @@ let send proxy_socket =
             in
             finally
               (fun () ->
-                let readable, _, _ = Unix.select [s_ip; s_unix] [] [] (-1.0) in
-                if List.mem s_unix readable then (
-                  let fd, _peer = Unix.accept s_unix in
-                  to_close := fd :: !to_close ;
-                  let buffer = Bytes.make (String.length token) '\000' in
-                  let n = Unix.recv fd buffer 0 (Bytes.length buffer) [] in
-                  let token' = Bytes.sub_string buffer 0 n in
-                  if token = token' then
-                    let (_ : int) =
-                      Fd_send_recv.send_fd_substring fd token 0
-                        (String.length token) [] proxy_socket
-                    in
-                    ()
-                ) else if List.mem s_ip readable then (
-                  let fd, _peer = Unix.accept s_ip in
-                  List.iter close !to_close ;
-                  to_close := fd :: !to_close ;
-                  proxy fd proxy_socket
-                ) else
-                  assert false
-                (* can never happen *)
+                let epoll = Polly.create () in
+                List.iter
+                  (fun fd -> Polly.add epoll fd Polly.Events.inp)
+                  [s_ip; s_unix] ;
+                Fun.protect
+                  ~finally:(fun () -> Polly.close epoll)
+                  (fun () ->
+                    ignore
+                    @@ Polly.wait epoll 2 (-1) (fun _ fd _ ->
+                           if s_unix = fd then (
+                             let fd, _peer = Unix.accept s_unix in
+                             to_close := fd :: !to_close ;
+                             let buffer =
+                               Bytes.make (String.length token) '\000'
+                             in
+                             let n =
+                               Unix.recv fd buffer 0 (Bytes.length buffer) []
+                             in
+                             let token' = Bytes.sub_string buffer 0 n in
+                             if token = token' then
+                               let (_ : int) =
+                                 Fd_send_recv.send_fd_substring fd token 0
+                                   (String.length token) [] proxy_socket
+                               in
+                               ()
+                           ) else if s_ip = fd then (
+                             let fd, _peer = Unix.accept s_ip in
+                             List.iter close !to_close ;
+                             to_close := fd :: !to_close ;
+                             proxy fd proxy_socket
+                           ) else
+                             assert false (* can never happen *)
```

Comment:
Wouldn't be so sure, we could get spurious notifications from 'epoll', and we don't want to crash the application (or thread). Perhaps just log it as error for now?
Especially that you've added 2 file descriptors s_ip and s_unix, this will NOT be just s_unix, it is entirely possible that s_ip will have something available, otherwise why add it to epoll in the first place?